File: backend/app/schemas/prices.py
"""Schemas — Price Intelligence Engine v2."""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field


class PriceProductResponse(BaseModel):
    id: UUID
    supplier: str | None
    manufacturer: str | None
    brand: str | None
    sku: str | None
    mpn: str | None
    model: str | None
    description: str | None
    category: str | None
    product_type: str
    excluded_from_laptop: bool = True
    is_cotizable: bool = False
    price_review_status: str = "ok"
    classification_label: str = "producto general"
    stock_source_column: str | None = None
    processor: str | None
    ram_gb: int | None
    storage_gb: int | None
    storage_type: str | None
    display: str | None
    operating_system: str | None
    price: Decimal | None
    preferred_price: Decimal | None = None
    preferred_price_field: str | None = None
    price_regular: Decimal | None = None
    price_rebate: Decimal | None = None
    price_discount: Decimal | None = None
    prices_original: dict = Field(default_factory=dict)
    currency: str
    stock: int | None
    in_transit: int | None
    stock_text_original: str | None = None
    warranty: str | None
    source_filename: str
    source_sheet: str | None
    source_row: int | None
    source_file_date: datetime | None = None
    source_file_date_estimated: bool = True
    file_id: UUID
    indexed_at: datetime

    model_config = {"from_attributes": True}


class PriceProductDetailResponse(PriceProductResponse):
    raw_row_json: dict = Field(default_factory=dict)
    raw_columns_json: list = Field(default_factory=list)
    comparison_price_label: str = "Precio usado para comparación"
    cotizable_warning: str | None = None


class PriceSearchResponse(BaseModel):
    query: str | None = None
    total: int
    items: list[PriceProductResponse]


class PriceCompareAlternative(BaseModel):
    supplier: str | None
    brand: str | None
    description: str | None
    sku: str | None
    price: Decimal | None
    preferred_price: Decimal | None = None
    currency: str
    stock: int | None
    source_filename: str
    source_sheet: str | None
    source_row: int | None = None
    file_date: datetime | None
    file_date_estimated: bool = True


class PriceCompareResponse(BaseModel):
    question: str
    best_supplier: str | None
    best_product: PriceProductResponse | None
    alternatives: list[PriceCompareAlternative] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    summary: str


class PriceSyncResponse(BaseModel):
    files_detected: int
    files_new: int
    files_unchanged: int
    files_updated: int
    records_created: int
    duration_ms: int
    errors: list[str] = Field(default_factory=list)


class PriceQuoteDraftCreate(BaseModel):
    product_id: UUID
    client_name: str | None = None
    quantity: int = Field(default=1, ge=1)
    margin_percent: Decimal | None = Field(default=Decimal("15"), ge=0, le=500)


class PriceQuoteDraftResponse(BaseModel):
    id: UUID
    product_id: UUID
    task_id: UUID | None = None
    client_name: str | None
    quantity: int
    description: str | None
    cost_price: Decimal | None
    currency: str
    supplier: str | None
    margin_percent: Decimal | None
    sale_price_suggested: Decimal | None
    source_filename: str | None
    source_sheet: str | None
    source_row: int | None
    source_file_date: datetime | None
    status: str
    odoo_product_id: int | None
    odoo_match_status: str | None
    user_id: UUID | None = None
    copy_line: str | None = None

    model_config = {"from_attributes": True}


class PriceQuoteDraftListResponse(BaseModel):
    items: list[PriceQuoteDraftResponse]
    total: int


class OdooProductMatchResponse(BaseModel):
    found: bool
    product_id: int | None = None
    product_name: str | None = None
    message: str
    action: str = "none"
    odoo_default_code: str | None = None

File: frontend/src/lib/prices.ts
/** Tipos — Price Intelligence Engine v2 */

export interface PriceProduct {
  id: string;
  supplier: string | null;
  manufacturer: string | null;
  brand: string | null;
  sku: string | null;
  mpn: string | null;
  model: string | null;
  description: string | null;
  category: string | null;
  product_type: string;
  excluded_from_laptop?: boolean;
  is_cotizable?: boolean;
  price_review_status?: string;
  classification_label?: string;
  stock_source_column?: string | null;
  processor: string | null;
  ram_gb: number | null;
  storage_gb: number | null;
  storage_type: string | null;
  display: string | null;
  operating_system: string | null;
  price: string | null;
  preferred_price: string | null;
  preferred_price_field: string | null;
  price_regular: string | null;
  price_rebate: string | null;
  price_discount: string | null;
  prices_original: Record<string, unknown>;
  currency: string;
  stock: number | null;
  in_transit: number | null;
  stock_text_original: string | null;
  warranty: string | null;
  source_filename: string;
  source_sheet: string | null;
  source_row: number | null;
  source_file_date: string | null;
  source_file_date_estimated: boolean;
  file_id: string;
  indexed_at: string;
}

export interface PriceProductDetail extends PriceProduct {
  raw_row_json: Record<string, unknown>;
  raw_columns_json: Array<{ column: string; index: number; value: unknown }>;
  comparison_price_label: string;
  cotizable_warning?: string | null;
}

export interface PriceSearchResponse {
  query: string | null;
  total: number;
  items: PriceProduct[];
}

export interface PriceCompareAlternative {
  supplier: string | null;
  brand: string | null;
  description: string | null;
  sku: string | null;
  price: string | null;
  preferred_price: string | null;
  currency: string;
  stock: number | null;
  source_filename: string;
  source_sheet: string | null;
  source_row: number | null;
  file_date: string | null;
  file_date_estimated: boolean;
}

export interface PriceCompareResponse {
  question: string;
  best_supplier: string | null;
  best_product: PriceProduct | null;
  alternatives: PriceCompareAlternative[];
  warnings: string[];
  summary: string;
}

export interface PriceSearchParams {
  q?: string;
  marca?: string;
  proveedor?: string;
  fabricante?: string;
  categoria?: string;
  tipo_producto?: string;
  ram_gb?: number;
  almacenamiento_gb?: number;
  procesador?: string;
  pantalla?: string;
  sistema_operativo?: string;
  precio_min?: number;
  precio_max?: number;
  moneda?: string;
  stock_min?: number;
  stock_disponible?: boolean;
  solo_disponibles?: boolean;
  limit?: number;
}

export interface PriceQuoteDraft {
  id: string;
  product_id: string;
  task_id?: string | null;
  client_name: string | null;
  quantity: number;
  description: string | null;
  cost_price: string | null;
  currency: string;
  supplier: string | null;
  margin_percent: string | null;
  sale_price_suggested: string | null;
  source_filename: string | null;
  source_sheet: string | null;
  source_row: number | null;
  source_file_date: string | null;
  status: string;
  copy_line: string | null;
}

export interface PriceOdooMatch {
  found: boolean;
  product_id?: number | null;
  product_name?: string | null;
  message: string;
  action?: string;
  odoo_default_code?: string | null;
}
